PHPLIB-1163 Create tutorial for using MongoDB with Bref (#1273)
mongodb/mongo-php-library: docs/tutorial/lambda-aws.txt

============================================

> Deploy to Serverless AWS Lambda with BrefPHP

Reviewer: "Serverless" seems unnecessary, at least as far as the page title goes. You end up clarifying this in the first paragraph.

```suggestion
Deploy to AWS Lambda with BrefPHP
```

Note: you named the file "lambda-aws" but refer to "AWS Lambda" here. Since the page title is incorporated into the URL, should they be consistent?

> Deploy to Serverless AWS Lambda with BrefPHP

Author: Also using `Bref` as your other comments.
```suggestion
Deploy to AWS Lambda with Bref
```

============================================
Deploy to Serverless AWS Lambda with BrefPHP
============================================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. versionadded:: 1.17

Overview
--------

`Bref <https://bref.sh>` allows to deploy serverless PHP applications on AWS Lambda.
In this tutorial, you will deploy a simple PHP application with the MongoDB PHP extension,
and connect to an Atlas cluster using AWS IAM authentication.

> Pre-requisites

Reviewer: ```suggestion
Prerequisites
```

--------------

> Before you begin, you must install bref on your machine. You can follow the

Reviewer: ```suggestion
Before you begin, you must install Bref on your machine. You can follow the
```

Bref's own docs capitalize the name, so I think it makes sense to do so throughout our own docs.

`official documentation to setup bref <https://bref.sh/docs/setup>`_.

> Install the mongodb extension

Reviewer: ```suggestion
Install the mongodb Extension
```

Optionally decide if you want to capitalize "MongoDB", but all lowercase is suitable if you intend to refer to the PECL package. Regardless, if this is a section title I think title capitalization rules apply.

Author: ```suggestion
Install the MongoDB extension
```

-----------------------------

By default, the bref layer is compiled with PHP and a few extensions. Additional extensions
are provided in additional layers.

Start by creating a new directory for your project and install the required mongodb
and bref dependencies. This project will be a bare minimum PHP web application that

> connects to a MongoDB cluster, no framework is used, but you can adapt the code to

Reviewer: ```suggestion
connects to a MongoDB cluster. No framework is used, but you can adapt the code to
```

This reads like it should be separate sentences.

your needs.

.. code-block:: bash
    mkdir bref-mongodb-app && cd bref-mongodb-app
    composer init
    composer require bref/bref bref/extra-php-extensions
    vendor/bin/bref init

> A very simple web page have been created. To validate the deployment, you can start

Reviewer: ```suggestion
A very simple web page has been created. To validate the deployment, you can start
```

Note: docs guidelines may want to avoid using subjective terms like "very simple", but I'll defer to the docs team to revise the language.

Reviewer 2: Regardless of docs team guidelines, I prefer avoiding such terms, especially in documentation aimed at beginners.

Author: ```suggestion
The file ``index.php`` have been created. To validate the deployment, you can start
```

Author: ```suggestion
The file ``index.php`` has been created. To validate the deployment, you can start
```

by deploying this default application.

.. code-block:: bash
    serverless deploy

> BrefPHP provides a Lambda layer with PHP and some very common extensions.

Reviewer: Note that you're referring to "BrefPHP" here but "bref" elsewhere. I think it should be consistent, and "Bref" is likely the right term.

Author: ```suggestion
Bref provides a Lambda layer with PHP and some very common extensions.
```

Additional extensions are provided by the package `bref/extra-php-extension <https://github.com/brefphp/extra-php-extensions>`.

.. code-block:: yaml

    plugins:
        - ./vendor/bref/bref
        - ./vendor/bref/extra-php-extensions

    functions:
        api:
            handler: index.php
            runtime: php-83-fpm
            layers:
                - ${bref-extra:mongodb-php-83}

Let's try to use the MongoDB driver with this simple web page that list planets

> from the `sample dataset <https://www.mongodb.com/docs/atlas/sample-data/>`.

Author: I need to fix the doc link.

Reviewer: Since this is RST syntax, shouldn't you have a double underscore after the link for an anonymous reference? For example:

```
`sample dataset <https://www.mongodb.com/docs/atlas/sample-data/>`__
```

Author: I think there is a prefix for links to the docs.

Replace the contents of ``index.php`` with the following:

.. literalinclude:: /examples/lambda-aws/index.php
   :language: php

Deploy the application

.. code-block:: bash
    serverless deploy

> The application will not work unless you provide a working ``MONGODB_URI`` variable.

Reviewer: ```suggestion
The application will not work unless you define the ``MONGODB_URI`` environment variable.
```

AWS Credentials
---------------

Atlas supports passwordless authentication with AWS credentials. In any Lambda function,
AWS sets environment variables that contains the access token and secret token with
the role assigned to deployed function.

> Set Up `unified AWS Access <https://www.mongodb.com/docs/atlas/security/set-up-unified-aws-access/>`

Reviewer: ```suggestion
Set up `unified AWS Access <https://www.mongodb.com/docs/atlas/security/set-up-unified-aws-access/>`
```

> 1. Open the lambda function in the AWS console

Reviewer: Earlier, "Lambda" was capitalized. There are some occurrences here and below that are lowercase.

Author: ```suggestion
1. Open the Lambda function in the AWS console
```

2. In "Configuration > Permission", copy the "Role name"
3. Open the MongoDB Atlas project
4. Go to "Security > Database Access"
5. Click "Add a new Database User"
6. Select Authentication Method: "AWS IAM", type "IAM Role" and paste the role name in "AWS Role ARN".
7. Add "Built-in Role": "Read and write any database"
8. Validate by clicking on "Add user".

Now that the permissions have been configured, the lambda function is allowed to access
your Atlas cluster. You can configure your application with the Atlas endpoint.

Update the ``serverless.yml`` file to pass the environment variable ``MONGODB_URI``

.. code-block:: yaml

    provider:
        environment:

>             MONGODB_URI: "mongodb+srv://cluster0.abcdefg.mongodb.net/"

Reviewer: ```suggestion
            MONGODB_URI: "mongodb+srv://cluster0.example.mongodb.net/"
```

This may make it more clear that it's a placeholder value. Feel free to disregard, though.
